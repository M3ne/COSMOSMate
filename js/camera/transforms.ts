import { DEG2RAD, RADIUS_OF_EARTH, TAU, ZOOM_EXP } from '../lib/constants';
import { getDayOfYear } from '../timeManager/transforms';

export const normalizeAngle = (angle: number): number => {
  angle %= TAU;
  if (angle > TAU / 2) angle -= TAU;
  if (angle < -TAU / 2) angle += TAU;
  return angle;
};
export const lon2yaw = (long: number, selectedDate: Date): number => {
  const realTime = new Date();
  let propTime = new Date();
  let angle = 0;

  // NOTE: camera formula sometimes is incorrect, but has been stable for over a year
  // NOTE: Looks wrong again as of 8/29/2020 - time of year issue?
  // NOTE: Could camera be related to daylight savings time? Subtracting one hour from selected date works
  const doy = getDayOfYear(selectedDate);
  const modifier = 1000 * 60 * 60 * (-11.23 + 0.065666667 * doy);

  propTime.setUTCHours(selectedDate.getUTCHours()); // + (selectedDate.getUTCMonth() * 2 - 11) / 2); // Offset has to account for time of year. Add 2 Hours per month into the year starting at -12.
  propTime.setUTCMinutes(selectedDate.getUTCMinutes());
  propTime.setUTCSeconds(selectedDate.getUTCSeconds());
  propTime = new Date(propTime.getTime() * 1 + modifier);

  realTime.setUTCHours(0);
  realTime.setUTCMinutes(0);
  realTime.setUTCSeconds(0);
  let longOffset = (propTime.getTime() - realTime.getTime()) / 60 / 60 / 1000; // In Hours
  if (longOffset > 24) longOffset = longOffset - 24;
  longOffset = longOffset * 15; // 15 Degress Per Hour longitude Offset

  angle = (long + longOffset) * DEG2RAD;
  angle = normalizeAngle(angle);
  return angle;
};
export const lat2pitch = (lat: number): number => {
  let pitch = lat * DEG2RAD;
  if (pitch > TAU / 4) pitch = TAU / 4; // Max 90 Degrees
  if (pitch < -TAU / 4) pitch = -TAU / 4; // Min -90 Degrees
  return pitch;
};
export const alt2zoom = (alt: number): number => {
  const distanceFromCenter = alt + RADIUS_OF_EARTH + 30;
  return Math.pow((distanceFromCenter - settingsManager.minZoomDistance) / (settingsManager.maxZoomDistance - settingsManager.minZoomDistance), 1 / ZOOM_EXP);
};
