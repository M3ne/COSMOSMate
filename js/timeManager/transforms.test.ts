import * as transforms from './transforms';
// @ponicode
describe('transforms.getDayOfYear', () => {
  test('0', () => {
    let param1: any = new Date('01-01-2030');
    let result: any = transforms.getDayOfYear(param1);
    expect(result).toMatchSnapshot();
  });

  test('1', () => {
    let param1: any = new Date('32-01-2020');
    let result: any = transforms.getDayOfYear(param1);
    expect(result).toMatchSnapshot();
  });

  test('2', () => {
    let param1: any = new Date('01-01-2020');
    let result: any = transforms.getDayOfYear(param1);
    expect(result).toMatchSnapshot();
  });

  test('3', () => {
    let param1: any = new Date('01-13-2020');
    let result: any = transforms.getDayOfYear(param1);
    expect(result).toMatchSnapshot();
  });

  test('4', () => {
    let param1: any = new Date('');
    let result: any = transforms.getDayOfYear(param1);
    expect(result).toMatchSnapshot();
  });
});

// @ponicode
describe('transforms.jday', () => {
  test('0', () => {
    let result: any = transforms.jday(368.0, 275.0, 3, -100, 4, 60.0);
    expect(result).toMatchSnapshot();
  });

  test('1', () => {
    let result: any = transforms.jday(366.0, 9, 28, 100, 15, 60.0);
    expect(result).toMatchSnapshot();
  });

  test('2', () => {
    let result: any = transforms.jday(367.0, 275, 4, -5.48, 29, 60.0);
    expect(result).toMatchSnapshot();
  });

  test('3', () => {
    let result: any = transforms.jday(367.0, 9.0, 4, 0, 0, 59.0);
    expect(result).toMatchSnapshot();
  });

  test('4', () => {
    let result: any = transforms.jday(367.0, 276, 3, 1, 3, 60.0);
    expect(result).toMatchSnapshot();
  });

  test('5', () => {
    transforms.jday();
  });
});

// @ponicode
describe('transforms.localToZulu', () => {
  test('0', () => {
    let param1: any = new Date('01-01-2030');
    transforms.localToZulu(param1);
  });

  test('1', () => {
    let param1: any = new Date('01-01-2020');
    transforms.localToZulu(param1);
  });

  test('2', () => {
    let param1: any = new Date('01-13-2020');
    transforms.localToZulu(param1);
  });

  test('3', () => {
    let param1: any = new Date('');
    transforms.localToZulu(param1);
  });
});

// @ponicode
describe('transforms.dateFromJday', () => {
  test('0', () => {
    transforms.dateFromJday(28, 1);
  });

  test('1', () => {
    transforms.dateFromJday(15, 28);
  });

  test('2', () => {
    transforms.dateFromJday(2, 15);
  });

  test('3', () => {
    transforms.dateFromJday(28, 2);
  });

  test('4', () => {
    transforms.dateFromJday(29, 15);
  });

  test('5', () => {
    transforms.dateFromJday(NaN, NaN);
  });
});

// @ponicode
describe('transforms.dateToLocalInIso', () => {
  test('0', () => {
    let param1: any = new Date('01-13-2020');
    transforms.dateToLocalInIso(param1);
  });

  test('1', () => {
    let param1: any = new Date('01-01-2030');
    transforms.dateToLocalInIso(param1);
  });

  test('2', () => {
    let param1: any = new Date('01-01-2020');
    transforms.dateToLocalInIso(param1);
  });
});
