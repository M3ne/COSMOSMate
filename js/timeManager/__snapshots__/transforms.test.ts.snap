// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transforms.getDayOfYear 0 1`] = `1`;

exports[`transforms.getDayOfYear 1 1`] = `NaN`;

exports[`transforms.getDayOfYear 2 1`] = `1`;

exports[`transforms.getDayOfYear 3 1`] = `13`;

exports[`transforms.getDayOfYear 4 1`] = `NaN`;

exports[`transforms.jday 0 1`] = `1863786.3368055555`;

exports[`transforms.jday 1 1`] = `1855000.677777778`;

exports[`transforms.jday 2 1`] = `1863426.2925`;

exports[`transforms.jday 3 1`] = `1855337.5006828704`;

exports[`transforms.jday 4 1`] = `1863456.5444444444`;
